### ISSUE TITLE
[Help Wanted] Use GloVe with concatenated features

### ISSUE THREAD

**Issue by davidstap** [OpenNMT/OpenNMT-py#1091]
Hi, first of all: thanks for this great library. 

I want to use GloVe embeddings concatenated with features: documents in my .src file consist of `w|f1|f2|f3|f4`, and I want `w` to be a GloVe embedding, whereas the four features should be concatenated. How do I do this?

When I create a vocabulary with `tools/embeddings_to_torch.py` given the `w|f1|f2|f3|f4` .src file, the keys look like `w|f1|f2|f3|f4` (e.g. `winning|VERB|amod|NONE|O`), but I would like the keys to be words (e.g. `winning`). Otherwise, it does not make sense to use GloVe embeddings since these `w|f1|f2|f3|f4` keys will not match with GloVe word vectors.

My current approach is as follows: I created a vocabulary using a .src file without features (only `w`, i.e. without features, but otherwise identical), and used the GloVe embeddings based on the vocabulary of this file. For training I use the `w|f1|f2|f3|f4` .src files and the `w` GloVe embeddings. However, I am not sure if this has the desired effect (probably not). During training, I use `-feat_merge concat` and `-feat_vec_exponent 0.7`.

**Comment by vince62s** [comment]
closing this, since the same question is on the forum.
Mayeb @pltrdy if you have to time to answer.

**Comment by pltrdy** [comment]
Indeed, the embedding to torch script does not consider features.

What you describe as your "current approach" seems a good workaround to me. I never really used features so there may be side effects, but I would guess that it would have the desired effect.

**Comment by pltrdy** [comment]
Not that, by default, pre-trained embedding will be updated during training (you can disable it)

**Comment by evasharma** [comment]
Thanks for the comments in this thread. I have a slightly different version of this problem stated above.
I don't have a pre-trained embedding instead want to train it and concatenate the embedding layer with one hot vector. But I want the one hot vector to remain unchanged while training. How can I modify the aforementioned current approach to achieve this?

**Comment by pltrdy** [comment]
@evasharma this would require very specific code to be added. You could try masking gradients in the embedding module backward function, this could work.

**Comment by evasharma** [comment]
@pltrdy Thanks for your reply. I have been trying to see if I can somehow read the one-hot vector in parallel and concatenate it with the output of the embedding layer. Wonder if that is possible? I am not sure if reading a feature separately like that is possible.
1. Could you explain more what you mean by "masking gradients in the embedding module backward function"?
2. Also could you point me to which function you are talking about?

**Comment by pltrdy** [comment]
Well, it depend what you mean by "possible". Without editing the code base, it's not, but with ad-hoc code, it should be. I'm not sure how relevant it is to concat one-hot vectors with dense embedding (just wondering, if you're sure, go for it). Anyway, this would require quite some effort since there's no similar mechanisms implemented: features are treated the same way in the code, there's no such mechanisms that bypass embeddings, you kinda have to "hack" the code.

Masking the gradient is not relevant (my bad), you better set `requires_grad=False` for your unchanging vector.

**Comment by evasharma** [comment]
@pltrdy Thanks again. So if I understand you correctly -  you mean that I can read the features the same way as mentioned above? Eg. Documents in my .src file should consist of format - w|f1|f2|f3|f4 where w is a word in the input article and f1, f2 ,f3, f4 is the one hot vector(remains the same for all the words in an input article). 
I understood that setting -feat_merge="concat" will concatenate the vector with the embedding, but since I want to update my word embeddings but not the feature vector, you said I should set the requires_grad to the result to not update it. I do not understand that part. 

In Embeddings.py : 
         vocab_sizes.extend(feat_vocab_sizes)
        emb_dims.extend(feat_dims)
        pad_indices.extend(feat_padding_idx)

        # The embedding matrix look-up tables. The first look-up table
        # is for words. Subsequent ones are for features, if any exist.
        emb_params = zip(vocab_sizes, emb_dims, pad_indices)
        embeddings = [nn.Embedding(vocab, dim, padding_idx=pad, sparse=sparse)
                      for vocab, dim, pad in emb_params]
        emb_luts = Elementwise(feat_merge, embeddings) .   <--- this is where the features are concatenated with the embeddings. Now I am not sure what you mean by "set the feature embedding to a 'number to one-hot vector' type of function, and set requires_grad to the result in order to not update it". Will the above mentioned code not handle concatenating the vector or do I have to write additional code for it? I thought I might just have to handle the part where only the embeddings get updated and my one hot vector doesn't.  Is it possible to set "requires_grad" to not update just the one-hot vector? If yes, which part of the code should I be looking at to make the required change?

**Comment by pltrdy** [comment]
The idea is to make use of existing code as much as possible. Currently, features, feature embedding, and feature aggregation (concatenation for instance) are implemented. 

Therefore I suggest:
1. define a single integer valued feature (`f1`) that hold the ID of the positive element of the one-hot vector, `w|f1` where `f1 = i` means a one-hot vector `v[j] = 1 if j = i, v[j] = 0 otherwise`.
2. use an 'embedding' function to represent that feature as a one-hot vector (then it's quite a 'fake' embedding, there's nothing to be learnt).
3. Concatenate features (done by default, nothing to do)

---

Let us define a module that represent integers as one-hot vectors:
```
class Onehot(nn.Module): 
    def __init__(self, dim): 
        super(Onehot, self).__init__() 
        self.dim = dim 
    def forward(self, input): 
        one_hot = torch.zeros(input.size() + (self.dim,)) 
        one_hot.scatter_(-1, input.unsqueeze(-1), 1)  
        return one_hot 
```

Then, we want to use it as our feature embedding, which is defined on [line 127 (onmt/modules/embeddings.py)](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/embeddings.py#L127).

I suggest replacing that line (currently, for experiment purpose) with:
```
# Manually define the embedding (experimental code, do not PR this)
vocab, dim, pad = vocab_sizes[0], emb_dims[0], pad_indices[0]
emb_f0 = nn.Embedding(vocab, dim, padding_idx=pad, sparse=sparse)
emb_f1 = Onehot(vocab_sizes[1])
embeddings = [emb_f0, emb_f1]

# left rest of the code unchanged
emb_luts = Elementwise(feat_merge, embeddings)
[...]
```

As you can see, our `Onehot` module does not have any parameters since it just transforms the input. This, part won't (cannot) be update by the gradient descent, by design.

I haven't tested it (just the `Onehot` module, independently).

Hope it helps.

**Comment by evasharma** [comment]
Thanks, this is very helpful. I will try this and get back to you.
